splunk-otel-android/src/main/java/com/splunk/rum/ErrorIdentifierExtractor.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright Splunk Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.splunk.rum;
+
+import android.app.Application;
+import android.content.Context;
+import android.content.pm.ApplicationInfo;
+import android.content.pm.PackageInfo;
+import android.content.pm.PackageManager;
+import android.os.Bundle;
+import android.util.Log;
+import androidx.annotation.Nullable;
+
+public class ErrorIdentifierExtractor {
+
+    private static final String SPLUNK_UUID_MANIFEST_KEY = "SPLUNK_OLLY_CUSTOM_UUID";
+    @Nullable private static ErrorIdentifierExtractor instance = null;
+
+    private final Context context;
+    private final PackageManager packageManager;
+    private final ApplicationInfo applicationInfo;
+
+    @Nullable private final String applicationId;
+    @Nullable private final String versionCode;
+    @Nullable private final String customUUID;
+
+    private ErrorIdentifierExtractor(Application application)
+            throws PackageManager.NameNotFoundException {
+        this.context = application.getApplicationContext();
+        this.packageManager = context.getPackageManager();
+        this.applicationInfo =
+                packageManager.getApplicationInfo(
+                        context.getPackageName(), PackageManager.GET_META_DATA);
+
+        this.applicationId = applicationInfo.packageName;
+        this.versionCode = retrieveVersionCode();
+        this.customUUID = retrieveCustomUUID();
+    }
+
+    @Nullable
+    public static synchronized ErrorIdentifierExtractor getInstance(Application application) {
```

Comment:
Prefer never to have `synchronized` on public methods. If it's really necessary to make this method thread safe, then let's use a private lock or concurrency classes.

I'd also question if this class really needs to be a singleton. It's probably only ever used at startup/creation time, and is the operation really that expensive?

I also think it's a bummer that this class has more than one responsibility -- it's both the data class that contains the 3 fields that matter (appId, code, uuid), but is also the thing that knows how to extract those from the application context. Might be nice to split these responsibilities apart.